rust/kernel/clk.rs

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Common clock framework.
+//!
+//! C header: [`include/linux/clk.h`](../../../../include/linux/clk.h)
+
+use crate::{bindings, device, error::Result, str::CStr, to_result};
+
+/// Represents `struct clk *`.
+///
+/// # Invariants
+///
+/// The pointer is valid.
+pub struct Clk(*mut bindings::clk);
+
+impl Clk {
+    /// Creates new clock structure from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be valid.
+    pub unsafe fn new(clk: *mut bindings::clk) -> Self {
+        Self(clk)
+    }
+
+    /// Lookups and obtains a managed reference to a clock producer.
+    ///
+    /// * `dev` - device for clock "consumer".
+    /// * `id` - ID of clock consumer.
+    pub fn get(dev: &impl device::RawDevice, id: Option<&CStr>) -> Result<Self> {
+        dev.clk_get(id)
+    }
+
+    /// Returns value of the rate field of `struct clk`.
+    pub fn get_rate(&self) -> usize {
+        // SAFETY: the pointer is valid by the type invariant.
+        unsafe { bindings::clk_get_rate(self.0) as usize }
+    }
+
+    /// Prepares and enables the underlying hardware clock.
+    ///
+    /// This function should not be called in atomic context.
+    pub fn prepare_enable(&mut self) -> Result {
+        // SAFETY: the pointer is valid by the type invariant.
+        to_result(|| unsafe { bindings::clk_prepare_enable(self.0) })
```

Comment:
Are successful calls to `clk_prepare_enable` expected to eventually be followed by `clk_disable_unprepare`? If so, I think here we need to return something that will trigger that call on drop (for example, `PreparedClk` or something like it); alternatively, we could have a flag in `Clk` that determines if the clock is enabled and therefore should be disabled on drop.

We need this sort of functionality to ensure that we have proper cleanup despite not using devm.

Reply:
Yes, it is definitely needed.

I will prepare a solution for it.
Thanks.

Reply:
@wedsonaf 

Can you please review the initial idea that I have?
The solution keeps the interface safe as shown in the example:

```rs
let clk = dev.clk_get(Some(c_str!("myclk")))?;
// clk.disable_unprepare(); // ERROR: method not found in `Clk`
let clk = clk.prepare_enable()?;
// let clk = clk.prepare_enable()?; // Cannot call twice, error: cannot move out of dereference of `EnabledClk`
let clk_yet_again: Clk = clk.disable_unprepare();
```

The implementation so far:
```rs
impl Clk {
    pub fn prepare_enable(self) -> Result<EnabledClk> {
        // SAFETY: the pointer is valid by the type invariant.
        to_result(|| unsafe { bindings::clk_prepare_enable(self.0) })?;
        Ok(EnabledClk(self))
    }
}
/// A clock variant that is prepared and enabled.
pub struct EnabledClk(Clk);

impl EnabledClk {
    /// Disables and later unprepares the underlying hardware clock prematurely.
    ///
    /// This function should not be called in atomic context.
    pub fn disable_unprepare(self) -> Clk {
        let mut clk = ManuallyDrop::new(self);
        // SAFETY: the pointer is valid by the type invariant.
        unsafe { bindings::clk_disable_unprepare(clk.0 .0) };
        core::mem::replace(&mut clk.0, Clk(core::ptr::null_mut()))
    }
}

impl Drop for EnabledClk {
    fn drop(&mut self) {
        // SAFETY: the pointer is valid by the type invariant.
        unsafe { bindings::clk_disable_unprepare(self.0 .0) };
    }
}

impl Deref for EnabledClk {
    type Target = Clk;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for EnabledClk {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
```

With a little bit of work this approach could handle such transition:

`Clk` -> `PreparedClk` -> `EnabledClk` -> `PreparedClk` -> `Clk`

What do you think?

Reply:
Very nice. I like this approach.

Reply:
Thanks! One thing that I think is an issue here is that this piece of code tries to emulate polymorphism using the `Deref` trait. For this patch, I may redefine the methods of the `Clk` for the sake of simplicity. Until the codebase remains short this shouldn't be an issue. The code reusability then may be addressed when the codebase would start growing.

Reply:
I think having stuck with `Deref` would have been better but, as you point out, the overlap ATM is small enough that it isn't an issue.